lib/puppetx/augeasproviders_grub/util.rb

```diff
@@ -152,7 +152,9 @@ def self.grub2_cfg_paths
         valid_paths = paths.map do |path|
           begin
             real_path = File.realpath(path)
-            real_path if (File.readable?(real_path) && !File.directory?(real_path))
+            real_path if (File.readable?(real_path) && !File.directory?(real_path)) &&
+               # Exclude stub files which include main config (e.g. Debian OS family)
+               (File.foreach(real_path).grep(/^configfile /).none?)
```

Comment:
This isn't a huge concern given the size of the files being processed, but this will read the entire file and process each line.

While it's not going to be one line, you may want to switch over to doing a looped read and breaking on the first match.

Reply:
> While it's not going to be one line, you may want to switch over to doing a looped read and breaking on the first match.

With my limited ruby knowledge, the best I can come up with is to change to this, which seems much less readable (but I'm sure there is a better way I am missing):
```ruby
valid_paths = paths.map do |path|
          begin
            real_path = File.realpath(path)
            if (File.readable?(real_path) && !File.directory?(real_path)) then
               # Exclude stub files which include main config (e.g. Debian OS family)
               File.foreach(real_path) do |line|
                  if line.match(/^configfile /) then
                     real_path = nil
                     break
                  end
               end
               real_path
            end
          rescue Errno::ENOENT
            nil
          end
        end.compact.uniq
```
Do you have a better implementation idea, or a construct in mind I could use instead of unrolling the expression this way?

Reply:
That looks about right honestly. The only changes that I can think of are:
1. You can remove the `begin/rescue` block if you move the assignment of `real_path` into the `if` statement since it has to exist and won't throw `ENOENT`
2. You can drop the `then` clauses (style)
3. You might want to change the space after `configfile` to `\s` in case someone drops a tab in there for whatever reason

So, something like (untested):

```ruby
valid_paths = paths.map do |path|
  real_path = nil

  if File.readable?(path) && !File.directory?(path)
    real_path = File.realpath(path)

    # Exclude stub files which include main config (e.g. Debian OS family)
    File.foreach(real_path) do |line|
      if line.match(/^configfile\s/)
        real_path = nil
        break
      end
    end
  end

  real_path
end.compact.uniq
```